README.md
Linux inotify for Erlang
========================

Inotify is on erlang port for the Linux inotify API allowing one to monitor
changes to files and directory in the filesystem.

* [API Documentation](https://hexdocs.pm/inotify/)
* [Rebar3 Installation (hex)](https://hex.pm/packages/inotify/)

Example
-------

    inotify_demo() ->
         inotify:start(x,y),
         TmpDir = inotify:watch("/tmp/", ?ALL),
         inotify:add_handler(TmpDir, ?MODULE, self()),

         file:open("/tmp/foo_bar_inotify_test", [read, write]),

         receive
             {[create],0,"foo_bar_inotify_test"} ->
                 io:format("Yeah!~n");
         end.


For the full example refer to the unit test:
https://github.com/sheyll/inotify/blob/master/test/inotify_test.erl


Release History
---------------

* 20180416 version 0.4.3 fix long file name handling
* 20180215 version 0.4.2 rebar3 and hex.pm support
* 2015???? version 0.4.1 don't build on windows
* 20130101 version 0.4.0 switch to rebar
* 20100206 version 0.3 on github
* 20090221 release 0.2 bug fix
* 20080929 initial release version 0.1


ChangeLog
2018-04-12 Version 0.4.3

	* Merge pull request #5 from **Neeraj** to fix handling of long paths

2018-02-15 Sven Heyll <sven.heyll@gmail.com>

        * Switch to rebar3

        * Add a Makefile

        * Support hex.pm

        * Update documentation

2013-01-01  Sven Heyll <sven.heyll@gmail.com>

        * Move src/test.erl to priv/test.erl

        * Remove auto tools support

        * Add rebar configuration

        * Add OTP Application

        * Add OTP gen_event

2010-02-01  Mats Cronqvist  <masse@kreditor.se>

	* c_src/inotify_erlang.c (note_remove):
	driver will always return either a 2-tuple (synchronously) or a
	5-tuple (asynchronously)

	* src/inotify.erl (test_start):
	see above

	* src/inoteefy.erl (talk_to_port):
	fixed race condition between synchronous and asynchronous events


test/inotify_test.erl
%%%=============================================================================
%%% @author Sven Heyll <sven.heyll@lindenbaum.eu>
%%% @copyright (C) 2013, Sven Heyll
%%%=============================================================================

-module(inotify_test).

-export([inotify_event/3]).

-include_lib("eunit/include/eunit.hrl").
-include("inotify.hrl").

%%%=============================================================================
%%% TESTS
%%%=============================================================================

monitor_file_test() ->
    process_flag(trap_exit, true),
    file:delete("/tmp/inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213"),

    inotify:start(x,y),
    Ref1 = inotify:watch("/tmp/", ?ALL),
    ok = inotify:add_handler(Ref1, ?MODULE, self()),
    receive after 100 -> ok end,

    {ok, H} = file:open("/tmp/inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213", [read, write]),

    receive
        {[create],0,"inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213"} -> pass;
        M1 -> throw({expected_message_file_create_but_got, M1})
    after 1000 -> throw(expected_create)
    end,
    receive
        {[open],0,"inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213"} -> pass;
        M2 -> throw({expected_message_file_open_but_got, M2})
    after 1000 -> throw(expected_open)
    end,
    Ref2 = inotify:watch("/tmp/inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213"),
    inotify:unwatch(Ref1),
    ok = inotify:add_handler(Ref2, ?MODULE, self()),
    receive after 100 -> ok end,
    io:write(H, "test"),
    receive
        {[modify],0,""} -> pass;
        M3 -> throw({expected_message_file_modify_but_got, M3})
    after 1000 -> throw(expected_modify)
    end,
    ok = file:close(H),
    receive
        {[close_write],0,""} -> pass;
        M4 -> throw({expected_message_file_close_write_but_got, M4})
    after 1000 -> throw(expected_close_write)
    end,
    Ref3 = inotify:watch("/tmp/", ?ALL),
    ok = inotify:add_handler(Ref3, ?MODULE, self()),
    receive after 100 -> ok end,
    ok = file:delete("/tmp/inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213"),
    receive
        {[attrib],0,""} -> pass
    after 1000 ->
            receive X1 -> throw({expected, X1})
            after 1000 -> throw(expected_attrib)
            end
    end,
    receive
        {[delete_self],0,""} -> pass
    after 1000 ->
            receive X2 -> throw({expected, X2})
            after 1000 -> throw(expected_delete_self)
            end
    end,
    receive
        {[ignored],0,""} -> pass
    after 1000 ->
            receive X3 -> throw({expected, X3})
            after 1000 -> throw(expected_ignored)
            end
    end,
    receive
        {[delete],0,"inotify_test22222222222222222222222222222222222222222222222222222222222222222222222221232132132131232132132132132132132132213"} -> pass
    after 1000 ->
            receive X4 -> throw({expected, X4})
            after 1000 -> throw(expected_delete)
            end
    end.

inotify_event(TestPid, Ref, Msg = ?inotify_msg(M, C, N)) ->
    io:format("$$$ ~p: ~p~n", [Ref, Msg]),
    TestPid ! {M, C, N},
    ok.


src/inotify.erl
%%%-------------------------------------------------------------------
%%% Created : 28 Jan 2010 by Mats Cronqvist <masse@kreditor.se>
%%% Converted to an OTP application by Sven Heyll <sven.heyll@gmail.com>
%%%
%%% @doc
%%% Application that monitors files via Linux `inotify'. A user can tell
%%% `inotify' to {@link watch/2} for specific events on several
%%% files/directories.  To actually get any events the User must call {@link
%%% add_handler/3}.
%%%
%%% To {@link unwatch/1} something the User must pass the return value of {@link
%%% watch/2} which is a unique reference to the monitor create by {@link
%%% watch/2}.
%%%
%%% This module also defines a behaviour, with a single function
%%% `inotify_event(Arg, EventRef, Msg)'
%%%
%%% A callback for file monitoring event handler added by {@link add_handler/3}.
%%%
%%% `Arg' is the user provided extra argument.
%%%
%%% `EventRef' is the value returned by {@link watch/2}.
%%%
%%% The `Msg' parameter of the callback function should be pattern-matched
%%% with the macro `?inotify_msg(Mask, Cookie, OptionalName)' contained
%%% in `include/inotify.hrl'.
%%%
%%% `Mask' is a list of atoms describing the nature of the event. Refer to the
%%% Linux inotify man page for a detailed description.
%%%
%%% `Cookie' is 0 except when a file is moved, where it is used to identify
%%% `move_from' and `move_to' events belonging to the same move-operation.
%%%
%%% `OptionalName' contains the relative file name when monitoring whole
%%% directories. E.g. when monitoring "/tmp" the following event might be
%%% created when opening "/tmp/xxx": `?inotify_msg([open], 0, "xxx")'.
%%% @end
%%%-------------------------------------------------------------------

-module(inotify).

-behaviour(application).
-behaviour(supervisor).

%% API
-export([watch/1, watch/2, unwatch/1, add_handler/3, print_events/1]).

%% Application callbacks
-export([start/2, stop/1]).

%% Supervisor callbacks
-export([init/1]).

%% inotify_evt callbacks
-export([inotify_event/3]).

-export_type([mask/0, msg/0]).

-define(SERVER, inotify_server).
-define(EVENT, inotify_evt).

-include("inotify.hrl").

-type msg() :: ?inotify_msg(Mask :: [mask()],
                            Cookie :: non_neg_integer(),
                            OptionalName :: string()).
%% A file monitoring message.

-type mask() :: ?ALL           |
                ?ACCESS        |
                ?ATTRIB        |
                ?CLOSE_WRITE   |
                ?CLOSE_NOWRITE |
                ?CLOSE         |
                ?CREATE        |
                ?DELETE        |
                ?DELETE_SELF   |
                ?MODIFY        |
                ?MOVE_SELF     |
                ?MOVED_FROM    |
                ?MOVED_TO      |
                ?MOVE          |
                ?OPEN          |
                ?DONT_FOLLOW   |
                ?MASK_ADD      |
                ?ONLYDIR.
%% The type of a file monitoring event.

%%%===================================================================
%%% Behaviour definition
%%%===================================================================

%%--------------------------------------------------------------------
%% FUCK YOU EDOC!
%%--------------------------------------------------------------------
-callback inotify_event(Arg :: term(),
                        EventRef :: term(),
                        Msg :: inotify:msg()) ->
    ok | remove_handler.

%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc
%% Add or modify a file/directory monitor for all file events. Events for the
%% file/directory will be published via a {@link inotify_evt}. The events
%% resulting from this call will be identified by the unique reference returned
%% by this function, no matter if the operation was successful or not.
%%
%% This reference must also be passed to {@link unwatch/1}.
%% @see inotify:add_handler/3
%% @see inotify:unwatch/1
%% @end
%%--------------------------------------------------------------------
-spec watch(string()) ->
                   reference().
watch(File) ->
    watch(File, ?ALL).

%%--------------------------------------------------------------------
%% @doc
%% Add or modify a file/directory monitor for specific events.  This
%% differs from {@link watch/1} by the additional `mask' argument.
%% Calling {@link watch/1} is equivalant to calling {@link watch/2}
%% with `?ALL' as second parameter.
%%
%% @see inotify:watch/1
%% @end
%%--------------------------------------------------------------------
-spec watch(string(), mask() | [mask()]) ->
                   reference().
watch(File, Mask) ->
    EventTag = make_ref(),
    ok = inotify_server:watch(File, EventTag, Mask),
    EventTag.

%%--------------------------------------------------------------------
%% @doc
%% Remove a monitor added via {@link watch/2}, `Ref' identifies the
%% monitor.
%% @end
%%--------------------------------------------------------------------
-spec unwatch(reference()) ->
                     ok.
unwatch(Ref) ->
    inotify_evt:remove_all_handlers(Ref),
    inotify_server:unwatch(Ref).

%%--------------------------------------------------------------------
%% @doc
%% Add an event handler that receives all events generated by a file monitor
%% identified by a tag that was passed to {@link inotify:watch/2}. When an event
%% occurs for `Ref' the handler will call `Module:inotify_event(Arg,
%% EventTag, Msg)' as defined by the `callback' definition in {@link
%% inotify_event/3}.
%% @end
%%--------------------------------------------------------------------
-spec add_handler(reference(), module(), term()) ->
                         ok.
add_handler(Ref, Module, Arg) ->
    inotify_evt:add_handler(Ref, Module, Arg).

%%--------------------------------------------------------------------
%% @doc
%% Add an event handler that prints to stdout events. The events to print are
%% identified by a reference returned by {@link watch/2}.
%% @end
%%--------------------------------------------------------------------
-spec print_events(reference()) ->
                          ok.
print_events(Ref) ->
    inotify_evt:add_handler(Ref, ?MODULE, []).


%%%===================================================================
%%% Application callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
start(_StartType, _StartArg) ->
    supervisor:start_link(?MODULE, []).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
stop(_) ->
    ok.

%%%===================================================================
%%% supervisor callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
init([]) ->
  {ok, {{one_for_all, 1, 1},
        [{?EVENT,  {?EVENT, start_link, []},
          permanent, 2000, worker, [?EVENT]},
         {?SERVER, {?SERVER, start_link, []},
          permanent, 2000, worker, [?SERVER]}]}}.

%%%===================================================================
%%% inotify_evt callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc
%% This module actually implements the {@link inotify_evt} behaviour, with a
%% function that simply prints the events to stdout.
%%
%% This function can be used as an example for how to implement a callback.
%%
%% @see print_events/1
%% @end
%%--------------------------------------------------------------------
inotify_event([], Ref, ?inotify_msg(Masks, Cookie, OptionalName)) ->
    io:format("[INOTIFY] - ~p - ~p ~p ~p~n", [Ref, Masks, Cookie, OptionalName]).


c_src/erl_comm.c
#include <unistd.h>
#include <stdint.h>
#include <stdlib.h>
#include <ei.h>
#include "erl_comm.h"

/*
 * inputs
 *    pbuf    a pointer to a pointer to the buffer to place the message in
 *    size    size of buf
 *    curpos  current offset into buf. Should be set to 0 before
 *            initial call
 *
 * returns  < 0 on read() error
 *         == 0 on read() returning 0
 *            1 when there is more to read
 *            2 when the message is complete
 */
int read_cmd(char **pbuf, int *size, int *curpos)
{
  int len;
  int count;
  int desired;

  char *buf = *pbuf;
  if (*curpos < 2) {
    /* read header */
    count = read(0, buf + *curpos, 2 - *curpos);

    if (count <= 0)
      return(count); /* Error or fd is closed */
    
    *curpos += count;
    if (*curpos < 2)
      return(1);
  }
  /* calculate the total message length and
   * the desired amount to read taking into account
   * the ammount already read
   */
  len = ((buf[0] << 8) & 0x0000ff00) | (buf[1] & 0x00ff);
  desired = len - *curpos + 2;

  /* check buffer size and realloc if necessary */
  if (len > *size) {
    char *newbuf = (char *) realloc(buf, len);
    if (newbuf == NULL)
      return -1;
    memset(*pbuf, 0, len - (*size));
    *pbuf = newbuf;
    buf = *pbuf;
    *size = len;
  }

  /* read message body */
  count = read(0, buf + *curpos, desired);
  if (count <= 0)
    return(0);

  *curpos += count;
  return(2);
}

int write_cmd(ei_x_buff *buff) {
  char li;

  li = (buff->index >> 8) & 0xff; 
  write_exact(&li, 1);
  li = buff->index & 0xff;
  write_exact(&li, 1);

  return write_exact(buff->buff, buff->index);
}

int write_exact(char *buf, int len) {
  int i, wrote = 0;

  do {
    if ((i = write(1, buf+wrote, len-wrote)) <= 0)
      return(i);
    wrote += i;
  } while (wrote<len);
 
  return(len);
}


c_src/inotify_driver.c
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/select.h>
#include <string.h>
#include <ei.h>
#include "inotify_erlang.h"
#include "erl_comm.h"

#define BUFFER_SIZE 100

int setup_select(note_t *nlist, fd_set *readfds) {
  int topfd, maxfd = 0;

  /* pselect initialisation */
  FD_ZERO(readfds);
  FD_SET(0, readfds);

  topfd = note_setup_select(nlist, readfds);
  if (topfd > maxfd)
    maxfd = topfd;

  return maxfd;
}

int process_command(const char *command, char *buf, int *index, note_t *nlist) {
  ei_x_buff  result;

  if (!strcmp("open", command)) {
    /* function open() */
    note_open(nlist, buf, index);
    return(0);
  }

  if (!strcmp("add", command)) {
    /* function int = add(int fd, char *path, int mask) */
    note_add(nlist, buf, index);
    return(0);
  }
  
  if (!strcmp("remove", command)) {
    /* function remove(fd, wd)  */
    note_remove(nlist, buf, index);
    return(0);
  }

  if (!strcmp("list", command)) {
    /* function list(nlist) */
    note_list(nlist, buf, index);
    return(0);
  }

  if (!strcmp("close", command)) {
    /* function close */
    note_close(nlist, buf, index);
    return(0);
  }

  /* Prepare the output buffer that will hold {ok, Result} or {error, Reason} */
  if (ei_x_new_with_version(&result) ||
      ei_x_encode_tuple_header(&result, 2) ||
      ei_x_encode_atom(&result, "error") ||
      ei_x_encode_atom(&result, "unsupported_command"))
    return(-1);

  write_cmd(&result);
  ei_x_free(&result);
  return(0);
} 

int main() {
  /* variables for erlang interface */
  int   index, version, arity;
  int   size = BUFFER_SIZE;
  int   cmdpos, result;
  char  *inbuf = NULL;
  char  command[MAXATOMLEN];

  /* varables for pselect */
  int maxfd, retval;
  fd_set readfds;

  /* inotify initialisation */
  note_t *nlist;
  if ((nlist = note_init()) == NULL) {
    perror("note_init()");
    exit(1);
  
  }

  inbuf = (char*) malloc(size);
  if (inbuf == NULL) {
      perror("malloc()");
      exit(1);
  }

  cmdpos = 0;
  maxfd = setup_select(nlist, &readfds);
  
  while ((retval = select(maxfd + 1, &readfds, NULL, NULL, NULL)) >= 0) {
    if FD_ISSET(0, &readfds) {
      memset(inbuf, 0, size);
      index = 0;
      result = read_cmd(&inbuf, &size, &cmdpos);

      if (result == 0) {
          free(inbuf);
          exit(1);
      } else if (result < 0) {
	/* exit(1); */
      } else if (result == 1) {
      } else {

	/* must add two(2) to inbuf pointer to skip message length header */
	if (ei_decode_version(inbuf+2, &index, &version) ||
	    ei_decode_tuple_header(inbuf+2, &index, &arity) ||
	    ei_decode_atom(inbuf+2, &index, command)) {
        free(inbuf);
	  exit(4);
	}

	process_command(command, inbuf+2, &index, nlist);

	/* reset position of inbuf */
	cmdpos = 0;
      }
    }
    /* check other file descriptors */
    note_read(nlist, &readfds);

    maxfd = setup_select(nlist, &readfds);
  }
  free(inbuf);
  exit(10);
}



src/inotify_server.erl
%%%-------------------------------------------------------------------
%%% Created : 28 Jan 2010 by Mats Cronqvist <masse@kreditor.se>
%%% Converted to an OTP application by Sven Heyll <sven.heyll@gmail.com>
%%%
%%% @doc
%%% A `gen_server' that provides access to low level Linux `inotify'.
%%% This is an internal module. Use {@link inotify} instead.
%%% @end
%%%-------------------------------------------------------------------

-module(inotify_server).

-behaviour(gen_server).

-author('Mats Cronqvist').
-author('Sven Heyll').

%% API
-export([start_link/0,
         watch/2,
         watch/3,
         unwatch/1]).

%% gen_server exports
-export([init/1,
         terminate/2,
         handle_cast/2,
         handle_info/2,
         handle_call/3,
         code_change/3]).

-define(SERVER, ?MODULE).

-include("inotify.hrl").

-record(ld,{port, fd}).

-define(log(T),
        error_logger:info_report(
          [process_info(self(), current_function),
           {line, ?LINE},
           T])).

%%%===================================================================
%%% API
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
start_link() ->
  gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
watch(File, EventTag) ->
    gen_server:call(?SERVER, {watch, {File, EventTag, ?ALL}}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
watch(File, EventTag, Mask) ->
    gen_server:call(?SERVER, {watch, {File, EventTag, Mask}}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
unwatch(EventTag) ->
    gen_server:cast(?SERVER, {unwatch, EventTag}).

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
init([]) ->
  process_flag(trap_exit, true),
  Port = open_port(),
  {ok, FD} = talk_to_port(Port , {open}),
  {ok, #ld{port = Port, fd = FD}}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
terminate(_, LD) ->
  talk_to_port(LD#ld.port, {close, LD#ld.fd}).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_info({Port, {data,Msg}}, LD = #ld{port = Port}) ->
  maybe_call_back(binary_to_term(Msg)),
  {noreply, LD};

handle_info(Msg, LD) ->
  ?log({unknown_message, Msg}),
  {noreply, LD}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_cast(stop, LD) ->
  {stop,normal, LD};

handle_cast({unwatch, Unwatch}, LD) ->
  {noreply, do_unwatch(Unwatch, LD)};

handle_cast(Msg, LD) ->
  ?log({unknown_message, Msg}),
  {noreply, LD}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
handle_call({watch,Watch},_From,LD) ->
    {Reply, NewLD} = do_watch(Watch, LD),
    {reply, Reply, NewLD}.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
code_change(_OldVsn, State, _Extra) ->
  {ok, State}.

%%%===================================================================
%%% Internal functions
%%%===================================================================

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
open_port() ->
    E = filename:join([code:priv_dir(inotify), "inotify"]),
    open_port({spawn, E},[{packet, 2}, binary, exit_status]).

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
maybe_call_back({event, WD, Mask, Cookie, Name}) ->
    case get({wd, WD}) of
        undefined ->
            case Mask of
                [ignored] ->
                    ok;
                _ ->
                    ?log([{got_event_without_callback, WD, Mask, Cookie, Name}])
            end;
        Tag ->
            inotify_evt:publish(Tag, ?inotify_msg(Mask, Cookie, Name))
    end.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
do_watch({File, Tag, Mask},LD) ->
    try
        {ok, WD} = talk_to_port(LD#ld.port, {add, LD#ld.fd, File, Mask}),
        put({tag, Tag}, WD),
        put({wd, WD}, Tag),
        {ok, LD}
    catch
        C:R ->
            ?log([{error_watching_file, File, Tag}, {C, R}]),
            {{error, C, R}, LD}
    end.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
do_unwatch(Tag, LD) ->
    case get({tag, Tag}) of
        undefined ->
            ?log([{not_watching, Tag}]),
            LD;
        WD ->
            try
                talk_to_port(LD#ld.port, {remove, LD#ld.fd, WD})
            catch
                C:R ->
                    ?log([{error_unwatching, Tag},{C, R}])
            end,
            erase({tag, Tag}),
            erase({wd, WD}),
            LD
    end.

%%--------------------------------------------------------------------
%% @private
%%--------------------------------------------------------------------
talk_to_port(Port,Msg) ->
    try
        erlang:port_command(Port, term_to_binary(Msg)),
        receive
            {Port, {data, D = <<131, 104, 2, _/binary>>}} ->
                binary_to_term(D)
        after 1000 ->
                throw(talk_to_port_timeout)
        end
    catch
        _:R ->
            throw({talking_to_port_failed, {R, Port, Msg}})
    end.
